File: Moedas/moedas.c
#include <stdio.h>
#include <stdlib.h>
#define MAX_N 100
#define MAX_M 100

void moedas(int n, int m, int V[], int T[MAX_M][MAX_N + 1]) {
    
    int i, j;
    //Preenchimento da 1ª linha
    for(j = 0; j <= n; j++) T[0][j] = 0;
    
    for(i = 0; i <= m; i++) T[i][0] = 1;
    
    for(i = 0; i <= m; i++) {
        for(j = 0; j <= n; j++) {
            if(j >= V[i]) {
                T[i][j] = T[i][j - V[i]] + T[i - 1][j];
            } else {
                T[i][j] = T[i - 1][j];
            }
        }
    }
}

int main() {
    int n, m, i, j;
    
    int V[MAX_M], T[MAX_M][MAX_N+1];
    
    // Leitura dos dados de entrada
    printf("Digite o valor a ser trocado: ");
    scanf("%d", &n);

    printf("Digite o numero de moedas: ");
    scanf("%d", &m);

    printf("Digite os valores das moedas:\n");
    
    for (i = 1; i <= m; i++) {
        scanf("%d", &V[i]);
    }
    
    // Preenchimento da tabela
    moedas(n, m, V, T);

    // Impressão da tabela
    printf("Tabela de troco:\n");
    for (i = 0; i <= m; i++) {
        for (j = 0; j <= n; j++) {
            printf("%d ", T[i][j]);
        }
        printf("\n");
    }

    return 0;
}

File: Moedas/moedas.py
def moedas(n, m, V):
    T = [[0 for _ in range(n + 1)] for _ in range(m + 1)]

    # Preenchimento da 1ª linha
    for j in range(n + 1):
        T[0][j] = 0

    for i in range(m + 1):
        T[i][0] = 1

    for i in range(1, m + 1):
        for j in range(1, n + 1):
            if j >= V[i]:
                T[i][j] = T[i][j - V[i]] + T[i - 1][j]
            else:
                T[i][j] = T[i - 1][j]

    return T


def main():
    n = int(input("Digite o valor a ser trocado: "))
    m = int(input("Digite o numero de moedas: "))

    V = [0] * (m + 1)

    print("Digite os valores das moedas:")
    for i in range(1, m + 1):
        V[i] = int(input())

    T = moedas(n, m, V)

    print("Tabela de troco:")
    for i in range(m + 1):
        for j in range(n + 1):
            print(T[i][j], end=" ")
        print()


if __name__ == "__main__":
    main()
